File: solutions/3909-minimum-increments-to-equalize-leaf-paths/minimum-increments-to-equalize-leaf-paths.java
class Solution {
    int operations = 0;

    public int minIncrease(int n, int[][] edges, int[] cost) {
        List<List<Integer>> adj = new ArrayList<>();
        for (int i = 0; i < n; i++) adj.add(new ArrayList<>());

        for (int[] edge : edges) {
            adj.get(edge[0]).add(edge[1]);
            adj.get(edge[1]).add(edge[0]);
        }

        dfs(0, -1, adj, cost);
        return operations;
    }

    private long dfs(int node, int parent, List<List<Integer>> adj, int[] cost) {
        long maxChild = 0;
        int count = 0, maxCount = 0;

        for (int child : adj.get(node)) {
            if (child == parent) continue;
            long childSum = dfs(child, node, adj, cost);
            if (childSum > maxChild) {
                maxChild = childSum;
                maxCount = 1;
            } else if (childSum == maxChild) {
                maxCount++;
            }
            count++;
        }

        operations += (count - maxCount);
        return cost[node] + maxChild;
    }
}

File: solutions/3909-minimum-increments-to-equalize-leaf-paths/minimum-increments-to-equalize-leaf-paths.cpp
class Solution {
public:
    // DFS returns max path sum from this node to a leaf
    long long dfs(int node, int parent, int& operations, vector<vector<int>>& adj, vector<int>& cost) {
        long long maxChildSum = 0;
        int childCount = 0, maxCount = 0;

        for (int neighbor : adj[node]) {
            if (neighbor == parent) continue;

            long long childSum = dfs(neighbor, node, operations, adj, cost);
            if (childSum > maxChildSum) {
                maxChildSum = childSum;
                maxCount = 1;
            } else if (childSum == maxChildSum) {
                maxCount++;
            }
            childCount++;
        }

        operations += (childCount - maxCount);
        return cost[node] + maxChildSum;
    }

    int minIncrease(int n, vector<vector<int>>& edges, vector<int>& cost) {
        vector<vector<int>> adj(n);
        for (auto& e : edges) {
            adj[e[0]].push_back(e[1]);
            adj[e[1]].push_back(e[0]);
        }

        int operations = 0;
        dfs(0, -1, operations, adj, cost);
        return operations;
    }
};